old/config.org
* Init.org
** Init, init?
*** Let there be sane defaults
**** Banishments
Send custom, backups, and autosaves to the shadow realm.
#+BEGIN_SRC emacs-lisp
  (setq custom-file "~/.emacs.d/custom.el")
  (load custom-file)
  (setq backups-directory-alist '(("." . "~/.emacs/backups")))
  (setq auto-save-file-name-transforms '((".*" "~/.emacs/auto-save-list/" t)))
#+END_SRC

Remove all scrollbars, tooltips, and other OS chrome.
#+BEGIN_SRC emacs-lisp
  (when window-system
    (tool-bar-mode -1)
    (scroll-bar-mode -1)
    (menu-bar-mode -1)
    (tooltip-mode -1))
#+END_SRC
*** Packages and Paths
#+BEGIN_SRC emacs-lisp
  (require 'package)
  (add-to-list 'package-archives
               '("melpa" . "https://melpa.org/packages/"))
  (package-initialize)
  (package-refresh-contents)
  (add-to-list 'load-path "~/.emacs.d/local-packages/")

  (require 'helpers)
#+END_SRC
** Appearances aren't everything, but they are /something/
*** Text rules
Wrap long lines
Maybe find that package that makes wraps prettier if they aren't yet?
#+BEGIN_SRC emacs-lisp
  (global-visual-line-mode t)
#+END_SRC
*** Font of, well, text
I like Maple Mono more than I've liked a font in a while
Ligatures are okay, I guess. Maybe I'll get sick of them.
#+BEGIN_SRC emacs-lisp
  (set-frame-font "Maple Mono 15" nil t)

  (use-package ligature
    :config
    (ligature-set-ligatures 't '("www"))
    (ligature-set-ligatures 'eww-mode '("ff" "fi" "ffi"))
    (ligature-set-ligatures
     'prog-mode
     '("|||>" "<|||" "<==>" "<!--" "####" "~~>" "***" "||=" "||>"
       ":::" "::=" "=:=" "===" "==>" "=!=" "=>>" "=<<" "=/=" "!=="
       "!!." ">=>" ">>=" ">>>" ">>-" ">->" "->>" "-->" "---" "-<<"
       "<~~" "<~>" "<*>" "<||" "<|>" "<$>" "<==" "<=>" "<=<" "<->"
       "<--" "<-<" "<<=" "<<-" "<<<" "<+>" "</>" "###" "#_(" "..<"
       "..." "+++" "/==" "///" "_|_" "www" "&&" "^=" "~~" "~@" "~="
       "~>" "~-" "**" "*>" "*/" "||" "|}" "|]" "|=" "|>" "|-" "{|"
       "[|" "]#" "::" ":=" ":>" ":<" "$>" "==" "=>" "!=" "!!" ">:"
       ">=" ">>" ">-" "-~" "-|" "->" "--" "-<" "<~" "<*" "<|" "<:"
       "<$" "<=" "<>" "<-" "<<" "<+" "</" "#{" "#[" "#:" "#=" "#!"
       "##" "#(" "#?" "#_" "%%" ".=" ".-" ".." ".?" "+>" "++" "?:"
       "?=" "?." "??" ";;" "/*" "/=" "/>" "//" "__" "~~" "(*" "*)"
       "\\\\" "://"))
    (global-ligature-mode t))
#+END_SRC
**** TODO Font install instructions
*** Size things up
Set width based on screen size, full height, and center
#+BEGIN_SRC emacs-lisp
  (let ((emacs-width
         (if (> (x-display-pixel-width) 1512) 180 120)))
    (progn
      (setq default-frame-alist
            '((left . 0.5)
              (fullscreen . fullheight)
              (user-position . t)))
      (add-to-list 'default-frame-alist (cons 'width emacs-width))))

#+END_SRC
*** Theme(s)
Still choosing between a few, so I randomized.
Other themes I didn't end up liking:
  base16-material-palenight base16-nebula base16-eris
Themes I maybe like:
  base16-deep-oceanic-next base16-moonlight base16-catppuccin-mocha
#+BEGIN_SRC emacs-lisp
  (use-package base16-theme :ensure t)
  (add-to-list 'custom-theme-load-path "~/.emacs.d/themes/")

  (let* ((themes
          '(base16-katy
            base16-eldritch
            tronesque
            weyland-yutani))
         (theme (random-element themes)))
    (progn (load-theme theme t)))
      ;(if (eq theme 'tronesque)
          ;(tronesque-mode-line))))
#+END_SRC
*** Make a splash
Custom splash screen. It could be better (a lot better);
#+BEGIN_SRC emacs-lisp
  (require 'splash-screen)
#+END_SRC
** Ivy
...and Counsel and Swiper
Ivy because it's lighter than Helm and more complete than ido.
Someday I should learn more about it and its capabilities.
#+BEGIN_SRC emacs-lisp
  (use-package ivy
    :ensure t
    :config
    (ivy-mode 1)
    (setq ivy-use-virtual-buffers t)
    (setq ivy-count-format ""))
#+END_SRC
** Evil
there's nothing lesser about it.
*** Setup Evil and Evil Collection
We don't use evil-org, because we have our own league of villians at home
#+BEGIN_SRC emacs-lisp
  ;; Evil
  (use-package evil
    :ensure t
    :init
    (setq evil-want-keybinding nil)
    :config
    (evil-mode 1))

  (use-package evil-collection
    :after evil
    :ensure t
    :config
    (evil-collection-init))

#+END_SRC
** Leadership Basics
Let's get down to business.
*** Setup the global leader, which-key for hinting, and our M-x replacement
#+BEGIN_SRC emacs-lisp
  (which-key-mode)

  (evil-set-leader 'motion (kbd "SPC"))

  (evil-define-key 'normal 'global (kbd "<leader>b" 'counsel-M-x))
#+END_SRC
*** Buffers
#+BEGIN_SRC emacs-lisp
  (evil-define-key 'normal 'global
    (kbd "<leader>b") (define-keymap
                        "b" #'counsel-switch-buffer
                        "l" #'list-buffers
                        "n" #'next-buffer
                        "p" #'previous-buffer))
#+END_SRC
*** Files & Directories
#+BEGIN_SRC emacs-lisp
  (evil-define-key 'normal 'global
    (kbd "<leader>d") (define-keymap "d" #'counsel-dired)
    (kbd "<leader>f") (define-keymap "f" #'counsel-find-file))
#+END_SRC
*** Projects
#+BEGIN_SRC emacs-lisp
  (evil-define-key 'normal 'global
    (kbd "<leader>p") (define-keymap
                        "b" #'project-switch-to-buffer
                        "f" #'project-find-file
                        "p" #'project-switch-project))
#+END_SRC
*** Lispy things
**** TODO Move to a local leader in elisp and similar modes and move to Language Modes > Lisp
#+BEGIN_SRC emacs-lisp
  (evil-define-key 'normal 'global
    (kbd "<leader>l") (define-keymap "l" #'eval-last-sexp))
#+END_SRC
*** Help!
I need somebody.
#+BEGIN_SRC emacs-lisp
  (evil-define-key 'normal 'global (kbd "<leader>h") help-map)
#+END_SRC
** Org
Org is love, org is life, org is all, org is eternal.
*** Basic org stuff
Setup basics including todo keywords and local leader binding.
I /don't/ use-package evil-org because I didn't agree with it when I tried it, but could revisit.
#+BEGIN_SRC emacs-lisp
  (require 'org)
  (setq org-default-notes-file "~/Code/todos.org")
  (setq org-todo-keywords
        '((sequence "TODO" "IN-PROGRESS" "WAITING" "DONE")
          (sequence "DRAFT(d)" "POST(p)" "|" "PUBLISH(B)")))

  (evil-define-key 'normal 'global
    (kbd "<leader>o") (define-keymap
                        "a" #'org-agenda
                        "c" #'org-capture))
  
  (with-eval-after-load 'org
    (evil-define-key 'normal org-mode-map
      ;(kbd "TAB") 'org-cycle          ; Fold/unfold headings
      ",t" 'org-todo                  ; Cycle TODO states
      ",d" 'org-deadline              ; Set deadline
      ",s" 'org-schedule              ; Schedule item
      ",p" 'org-priority              ; Set priority
      ",z" 'org-archive-subtree
      ">" 'org-shiftmetaright          ; Promote heading (** → ***)
      "<" 'org-shiftmetaleft))
#+END_SRC
*** Org Capture
#+BEGIN_SRC emacs-lisp
  (add-hook 'org-capture-mode-hook
            (lambda ()
              (evil-define-key 'normal 'local
                (kbd ",,") 'org-capture-finalize
                (kbd ",c") 'org-capture-finalize
                (kbd ",k") 'org-capture-kill
                (kbd ",r") 'org-capture-refile)))
  
  (setq org-capture-templates
        '(("b" "Blog Post" entry (file "~/Code/blog.org")
           "* DRAFT %i %?\n:PROPERTIES:\n ::EXPORT_FILE_NAME: index\n :TITLE: \n EXPORT_HUGO_DRAFT: true\n :EXPORT_HUGO_CUSTOM_FRONT_MATTER: :subtitle \n :EXPORT_HUGO_CUSTOM_FRONT_MATTER+: :summary \n:END:\n\n"
           :prepend t
           :empty-lines 1
           :jump-to-captured t)
          ("t" "Todo" entry (file+headline "~/Code/todos.org" "TODOs")
           "*** TODO %?\n%U\n")
          ("l" "Long Term" entry (file+headline "~/Code/todos.org" "Long Term")
           "*** TODO %?\n%U\n")
          ("p" "Project" entry (file+function "~/Code/todos.org" org-capture-select-project)
           "*** TODO %?\n%U\n")
          ("n" "Note" entry (file+olp+datetree "~/Code/notes.org")
           "**** %?\n")))
#+END_SRC
*** Org Agenda
#+BEGIN_SRC emacs-lisp
  (with-eval-after-load 'org-agenda
    (evil-define-key 'motion org-agenda-mode-map
      ",d" 'org-agenda-deadline
      ",s" 'org-agenda-schedule
      ",p" 'org-agenda-priority))

  (custom-set-variables
   '(org-agenda-files (quote ("~/Code/todos.org")))
   '(org-default-notes-file "~/Code/todos.org")
   '(org-agenda-span 7)
   '(org-agenda-start-on-weekday nil)
   '(org-deadline-warning-days 14)
   '(org-agenda-show-all-dates t)
   '(org-agenda-skip-deadline-if-done t)
   '(org-agenda-skip-scheduled-if-done t)
   '(org-reverse-note-order t)
   '(org-fast-tag-selection-single-key (quote expert))
   '(org-agenda-custom-commands
     (quote (("A" "Priority A Focus"
              ((agenda ""
                ((org-agenda-span 7)
                 (org-agenda-skip-function
                  (lambda ()
                    (org-agenda-skip-entry-if 'notregexp "\\[#A\\]")))
                 (org-agenda-overriding-header "Priority A Items This Week")))))
             ("P" "Priority A/B Todos"
              ((tags-todo "+PRIORITY=\"A\""
                ((org-agenda-overriding-header "Priority A Tasks:")))
               (tags-todo "+PRIORITY=\"B\""
                ((org-agenda-overriding-header "Priority B Tasks:")))))
             ("U" "Unscheduled Todos"
              ((alltodo ""
                ((org-agenda-skip-function
                  (lambda ()
                    (org-agenda-skip-entry-if 'scheduled 'deadline)))
                 (org-agenda-overriding-header "Unscheduled TODO entries:")))))))))
#+END_SRC
*** ox (org-export)
This lets us export to a hugo-powered blog.
#+BEGIN_SRC emacs-lisp
  (use-package ox-hugo
    :ensure t
    :pin melpa
    :after ox
    :config
    (setq org-hugo-base-dir "~/Code/blog/"))
#+END_SRC
**** TODO bind this to a local (or global org) leader key instead of C-c C-e or whatever it is
** Major Packages
*** Magit
...is the best
#+BEGIN_SRC emacs-lisp
  (use-package magit :ensure t)

  (evil-define-key 'normal 'global (kbd "<leader>g") 'magit)
#+END_SRC
*** Ag
The silver surfer. rg is faster, but whatevs, ag is fast enough.
TODO hook it up to more ivy stuff.
#+BEGIN_SRC emacs-lisp
  (use-package ag :ensure t)
  
  (evil-define-key 'normal 'global (kbd "<leader>s") (define-keymap "s" #'counsel-ag)
#+END_SRC
*** Vterm
#+BEGIN_SRC emacs-lisp
  (use-package vterm :ensure t)

  (use-package inheritenv
    :vc (:url "https://github.com/purcell/inheritenv" :rev :newest))

  (evil-define-key 'normal 'global
    (kbd "<leader>t") (define-keymap "t" 'vterm))
#+END_SRC
*** Claude
**** Monet
Monet lets us view Claude's edits as buffers in emacs and then y/q to accept/reject
It probably does other stuff. Figure out what that is.
#+BEGIN_SRC emacs-lisp
  (use-package monet
    :vc (:url "https://github.com/stevemolitor/monet" :rev :newest))
#+END_SRC
**** Claude Code
#+BEGIN_SRC emacs-lisp
  (use-package claude-code
    :vc (:url "https://github.com/stevemolitor/claude-code.el" :rev :newest)
    :config
    (setq claude-code-terminal-backend 'vterm)
    (add-hook
     'claude-code-process-environment-functions #'monet-start-server-function)
    (monet-mode 1)
    (claude-code-mode)
    :bind (:repeat-map my-claude-code-map ("M" . claude-code-cycle-mode)))
  (require 'claude-code)

  (evil-define-key 'normal 'global (kbd "<leader>c")  claude-code-command-map)
#+END_SRC
** Language Modes
*** The Basics (not BASIC!)
Universal or near-universal coding stuff
#+BEGIN_SRC emacs-lisp
  (setq-default tab-width 2)
  (setq custom-tab-width 2)
  (setq-default indent-tabs-mode nil)
#+END_SRC
*** CSS
#+BEGIN_SRC emacs-lisp
  (setq-default css-indent-offset 2)
#+END_SRC
*** HTML
#+BEGIN_SRC emacs-lisp
  (add-to-list 'auto-mode-alist '("\\.hbs\\'" . html-mode))
#+END_SRC
*** JavaScript
#+BEGIN_SRC emacs-lisp
  (setq-default js-indent-level custom-tab-width)

  (add-to-list 'auto-mode-alist '("\\.mjs\\'" . javascript-mode))
#+END_SRC
*** Lisp
#+BEGIN_SRC emacs-lisp
#+END_SRC
*** Markdown
#+BEGIN_SRC emacs-lisp
  (use-package markdown-mode)
#+END_SRC

** TODOs
*** TODO Diminish?
*** TODO More lispy local bindings
*** TODO ,, and ,k and ,' for editing code in org-babel


old/config.el
;; Packages
(require 'package)
(add-to-list 'package-archives
	     '("melpa" . "https://melpa.org/packages/"))
(package-initialize)
(package-refresh-contents)
(add-to-list 'load-path "~/.emacs.d/local-packages/")

;; Custom
(setq custom-file "~/.emacs.d/custom.el")
(load custom-file)

(require 'helpers)

;; Markdown
(use-package markdown-mode)

;; Magit
(use-package magit)

;; Ag
(use-package ag)

;; Ivy (and Counsel and Swiper)
(use-package ivy
  :ensure t
  :config
  (ivy-mode 1)
  (setq ivy-use-virtual-buffers t)
  (setq ivy-count-format ""))

;; Evil
(use-package evil
  :ensure t
  :init
  (setq evil-want-keybinding nil)
  :config
  (evil-mode 1))

(use-package evil-collection
  :after evil
  :ensure t
  :config
  (evil-collection-init))

(require 'org)

(add-hook 'org-capture-mode-hook
          (lambda ()
            (evil-define-key 'normal 'local
              (kbd ",,") 'org-capture-finalize
              (kbd ",c") 'org-capture-finalize
              (kbd ",k") 'org-capture-kill
              (kbd ",r") 'org-capture-refile)))

(setq org-todo-keywords
      '((sequence "TODO" "IN-PROGRESS" "WAITING" "DONE")))
(setq org-default-notes-file "~/Code/todos.org")
(setq org-capture-templates
      '(("b" "Blog Post" entry (file "~/Code/blog.org")
         "* DRAFT %i %?\n:PROPERTIES:\n ::EXPORT_FILE_NAME: index\n :TITLE: \n EXPORT_HUGO_DRAFT: true\n :EXPORT_HUGO_CUSTOM_FRONT_MATTER: :subtitle \n :EXPORT_HUGO_CUSTOM_FRONT_MATTER+: :summary \n:END:\n\n"
         :prepend t
         :empty-lines 1
         :jump-to-captured t)
        ("t" "Todo" entry (file+headline "~/Code/todos.org" "TODOs")
         "*** TODO %?\n%U\n")
        ("l" "Long Term" entry (file+headline "~/Code/todos.org" "Long Term")
         "*** TODO %?\n%U\n")
        ("p" "Project" entry (file+function "~/Code/todos.org" org-capture-select-project)
         "*** TODO %?\n%U\n")
        ("n" "Note" entry (file+olp+datetree "~/Code/notes.org")
         "**** %?\n")))

;; Org-agenda evil keybindings
(with-eval-after-load 'org-agenda
  (evil-define-key 'motion org-agenda-mode-map
    ",d" 'org-agenda-deadline
    ",s" 'org-agenda-schedule
    ",p" 'org-agenda-priority))

(custom-set-variables
 '(org-agenda-files (quote ("~/Code/todos.org")))
 '(org-default-notes-file "~/Code/todos.org")
 '(org-agenda-span 7)
 '(org-agenda-start-on-weekday nil)
 '(org-deadline-warning-days 14)
 '(org-agenda-show-all-dates t)
 '(org-agenda-skip-deadline-if-done t)
 '(org-agenda-skip-scheduled-if-done t)
 '(org-reverse-note-order t)
 '(org-fast-tag-selection-single-key (quote expert))
 '(org-agenda-custom-commands
   (quote (("A" "Priority A Focus"
            ((agenda ""
              ((org-agenda-span 7)
               (org-agenda-skip-function
                (lambda ()
                  (org-agenda-skip-entry-if 'notregexp "\\[#A\\]")))
               (org-agenda-overriding-header "Priority A Items This Week")))))
           ("P" "Priority A/B Todos"
            ((tags-todo "+PRIORITY=\"A\""
              ((org-agenda-overriding-header "Priority A Tasks:")))
             (tags-todo "+PRIORITY=\"B\""
              ((org-agenda-overriding-header "Priority B Tasks:")))))
           ("U" "Unscheduled Todos"
            ((alltodo ""
              ((org-agenda-skip-function
                (lambda ()
                  (org-agenda-skip-entry-if 'scheduled 'deadline)))
               (org-agenda-overriding-header "Unscheduled TODO entries:")))))))))

;; Org Blogging stuff
(add-to-list 'org-todo-keywords '(sequence "DRAFT(d)" "POST(p)" "|" "PUBLISH(B)"))

(use-package ox-hugo
  :ensure t
  :pin melpa
  :after ox
  :config
  (setq org-hugo-base-dir "~/Code/blog/"))

;(use-package evil-org
  ;"TBD if a good idea or make your own"
  ;...see-eg-version)

;; Claude
(use-package vterm)

(use-package inheritenv
  :vc (:url "https://github.com/purcell/inheritenv" :rev :newest))

(use-package monet
  :vc (:url "https://github.com/stevemolitor/monet" :rev :newest))

(use-package claude-code
  :vc (:url "https://github.com/stevemolitor/claude-code.el" :rev :newest)
  :config
  (setq claude-code-terminal-backend 'vterm)
  (add-hook
   'claude-code-process-environment-functions #'monet-start-server-function)
  (monet-mode 1)
  (claude-code-mode)
  :bind (:repeat-map my-claude-code-map ("M" . claude-code-cycle-mode)))
(require 'claude-code)

;; Keybindings
(which-key-mode)

;; Evil leader stuff
(evil-set-leader 'motion (kbd "SPC"))
(evil-define-key 'normal 'global
  (kbd "<leader>SPC") 'counsel-M-x
  (kbd "<leader>b") (define-keymap
                      "b" #'counsel-switch-buffer
                      "l" #'list-buffers
                      "n" #'next-buffer
                      "p" #'previous-buffer)
  (kbd "<leader>c")  claude-code-command-map
  (kbd "<leader>d") (define-keymap
                      "d" #'counsel-dired)
  (kbd "<leader>f") (define-keymap
                      "f" #'counsel-find-file)
  (kbd "<leader>g") 'magit
  (kbd "<leader>h") help-map
  (kbd "<leader>l") (define-keymap
                      "l" #'eval-last-sexp)
  (kbd "<leader>o") (define-keymap
                      "a" #'org-agenda
                      "c" #'org-capture)
  (kbd "<leader>p") (define-keymap
                      "b" #'project-switch-to-buffer
                      "f" #'project-find-file
                      "p" #'project-switch-project)
  (kbd "<leader>s") (define-keymap
                      "s" #'counsel-ag)
  (kbd "<leader>t") (define-keymap
                      "t" 'vterm))

;; Org-mode evil keybindings
(with-eval-after-load 'org
  (evil-define-key 'normal org-mode-map
    ;(kbd "TAB") 'org-cycle          ; Fold/unfold headings
    ",t" 'org-todo                  ; Cycle TODO states
    ",d" 'org-deadline              ; Set deadline
    ",s" 'org-schedule              ; Schedule item
    ",p" 'org-priority              ; Set priority
    ",z" 'org-archive-subtree
    ">" 'org-shiftmetaright          ; Promote heading (** → ***)
    "<" 'org-shiftmetaleft))

(setq-default tab-width 2)

; Javascript!
(setq-default indent-tabs-mode nil)
(setq custom-tab-width 2)
(setq-default js-indent-level custom-tab-width)

;; CSS
(setq-default css-indent-offset 2)

;; File extension mapping
(add-to-list 'auto-mode-alist '("\\.mjs\\'" . javascript-mode))
(add-to-list 'auto-mode-alist '("\\.hbs\\'" . html-mode))

;; Wrap text everywhere
(global-visual-line-mode t)
(tool-bar-mode -1)
(scroll-bar-mode -1)
(menu-bar-mode -1)

;; Splash
(require 'splash-screen)

;; Set width based on screen size, full height, and center
(let ((emacs-width
       (if (> (x-display-pixel-width) 1512) 180 120)))
  (progn
    (setq default-frame-alist
          '((left . 0.5)
            (fullscreen . fullheight)
            (user-position . t)))
    (add-to-list 'default-frame-alist (cons 'width emacs-width))))

;; Themes
(use-package base16-theme :ensure t)
(add-to-list 'custom-theme-load-path "~/.emacs.d/themes/")

(let* ((themes
        '(;base16-nebula
          ;base16-material-palenight
          base16-katy
          base16-eldritch
          ;maybe base16-deep-oceanic-next
          ;base16-eris
          ; maybe base16-moonlight
          ; maybe base16-catppuccin-mocha
          tronesque
          weyland-yutani))
       (theme (random-element themes)))
  (progn (load-theme theme t)))
    ;(if (eq theme 'tronesque)
        ;(tronesque-mode-line))))

;; Font Tweaks
(set-frame-font "Maple Mono 15" nil t)

(use-package ligature
  :config
  (ligature-set-ligatures 't '("www"))
  (ligature-set-ligatures 'eww-mode '("ff" "fi" "ffi"))
  (ligature-set-ligatures
   'prog-mode
   '("|||>" "<|||" "<==>" "<!--" "####" "~~>" "***" "||=" "||>"
     ":::" "::=" "=:=" "===" "==>" "=!=" "=>>" "=<<" "=/=" "!=="
     "!!." ">=>" ">>=" ">>>" ">>-" ">->" "->>" "-->" "---" "-<<"
     "<~~" "<~>" "<*>" "<||" "<|>" "<$>" "<==" "<=>" "<=<" "<->"
     "<--" "<-<" "<<=" "<<-" "<<<" "<+>" "</>" "###" "#_(" "..<"
     "..." "+++" "/==" "///" "_|_" "www" "&&" "^=" "~~" "~@" "~="
     "~>" "~-" "**" "*>" "*/" "||" "|}" "|]" "|=" "|>" "|-" "{|"
     "[|" "]#" "::" ":=" ":>" ":<" "$>" "==" "=>" "!=" "!!" ">:"
     ">=" ">>" ">-" "-~" "-|" "->" "--" "-<" "<~" "<*" "<|" "<:"
     "<$" "<=" "<>" "<-" "<<" "<+" "</" "#{" "#[" "#:" "#=" "#!"
     "##" "#(" "#?" "#_" "%%" ".=" ".-" ".." ".?" "+>" "++" "?:"
     "?=" "?." "??" ";;" "/*" "/=" "/>" "//" "__" "~~" "(*" "*)"
     "\\\\" "://"))
  (global-ligature-mode t))
